subsys/bluetooth/host/gatt.c

```diff
@@ -3884,47 +3985,69 @@ static void gatt_prepare_write_rsp(struct bt_conn *conn, uint8_t err,
 	bt_gatt_write(conn, params);
 }
 
-static int gatt_prepare_write(struct bt_conn *conn,
-			      struct bt_gatt_write_params *params)
+static int gatt_prepare_write_buf(struct net_buf *buf, size_t len,
+				  void *user_data)
 {
-	struct net_buf *buf;
+	struct bt_gatt_write_params *params = user_data;
 	struct bt_att_prepare_write_req *req;
-	uint16_t len;
+	uint8_t *data;
 	size_t write;
 
-	len = MIN(params->length, bt_att_get_mtu(conn) - sizeof(*req) - 1);
+	req = net_buf_add(buf, sizeof(*req));
+	req->handle = sys_cpu_to_le16(params->handle);
+	req->offset = sys_cpu_to_le16(params->offset);
 
-	buf = bt_att_create_pdu(conn, BT_ATT_OP_PREPARE_WRITE_REQ,
-				sizeof(*req) + len);
-	if (!buf) {
+	data = ((uint8_t *)params->data) + params->offset;
+	write = net_buf_append_bytes(buf, len - sizeof(*req), data,
```

Comment:
I can't quite understand why are you skipping the offset? The original code used params->data directly, the offset is meant for the server where to start writing not for the client.

Reply:
The reason why I changed this code was because data pointer should not be modified, otherwise the encode function cannot properly re-encode it when resending.
But you are right, my change would not work properly in the case where the application has set an offset.

I have fixed this now, the data pointer is manipulated, as well as the length and offset like it was in the original code, except it is now  done once the prepare_write_rsp has been received.

There was also an off-by one error, and the EDTT tests did not catch this, I have tested it manually with the shell now.
Also added a fix to the shell for when the hex convert failed.